lib/client.js

```diff
@@ -143,6 +156,58 @@ extend(Raven.prototype, {
     return this;
   },
 
+  makeErrorHandler: function () {
+    var self = this;
+    var caughtFirstError = false;
+    var caughtSecondError = false;
+    var calledFatalError = false;
+    var firstError;
+    return function (err) {
+      if (!caughtFirstError) {
+        // this is the first uncaught error and the ultimate reason for shutting down
+        // we want to do absolutely everything possible to ensure it gets captured
+        // also we want to make sure we don't go recursion crazy if more errors happen after this one
+        firstError = err;
+        caughtFirstError = true;
+        self.captureException(err, function (sendErr, eventId) {
+          if (!calledFatalError) {
+            calledFatalError = true;
+            self.onFatalError(err, sendErr, eventId);
+          }
+        });
+      } else if (calledFatalError) {
+        // we hit an error *after* calling onFatalError - pretty boned at this point, just shut it down
+        utils.consoleAlert('uncaught exception after calling fatal error shutdown callback - this is bad! forcing shutdown');
```

Comment:
Yea, if there's another independent error after onFatalError starts running we're just gonna shut down, but I'm not sure if we should, or even *can* reliably, worry about ensuring the completion of onFatalError - there's no way for us to really know how long onFatalError should take, whereas we do have some rough idea of how long capturing the first error should take. I'm much more worried about ensuring that we successfully capture the original first error and that the process shuts down.

It could make sense to expose some sort of user option for deciding where we should sit on the bail-at-first-trouble vs wait-it-out-and-see spectrum; right now as long as fatalError has been called, we bail if any further bad stuff happens, mostly in order to ensure the process exits and doesn't hang. I'm not sure what that option would look like or what the use case for it would be, so I kind of want to defer it; anything being done in the onFatalError callback is a best-effort no-guarantees sort of thing anyway.